### ISSUE TITLE
Distressing proliferation of ambient authority

### ISSUE THREAD

**Issue by erights** [Agoric/agoric-sdk#2160]
A simple search

![Screen Shot 2021-01-02 at 12 00 38 AM](https://user-images.githubusercontent.com/273868/103453343-e0c46880-4c8d-11eb-91c3-97c8998e95b9.png)

shows 26 modules (!!) in agoric-sdk importing `from 'fs';` after filtering out tests and demo code. What other sources of ambient authority are we using? How should we detect/prevent/warn on these by default? How should we override that default for the (should be) ***rare*** resource module that should directly access ambient authority?

Attn @dckc

**Comment by dckc** [comment]
I'd like to see "do not export powerful objects" on a code review checklist. It's one thing to import `fs` in a rollup config; that's more or less unavoidable. But it's quite another for packages / modules that we develop to export objects that close over `fs`.

Do we have a code review checklist? By way of example: https://informatics.kumc.edu/work/wiki/CodeReviewNotes

For cleaning up existing code, lavamoat would let us detect use of ambient authority in many / most cases, I suppose.

Strictly speaking, I don't think this constitutes a **bug**, since I don't see any incorrect observable behavior identified. In KUMC Informatics, we would label this as **technical-debt** and review the list of such issues every few weeks in a QA meeting. Making time to pay off technical debt in the schedule when it competes with observable bugs and features is a challenge, but one that healthy teams eventually deal with.

**Comment by katelynsills** [comment]
For what it is worth, the deploy scripts are taking local files and bundling and deploying them, so they must have the ability to read parts of the filesystem.

**Comment by dckc** [comment]
Yes, but shouldn't that access to the filesystem be passed in explicitly? I don't see any particular reason why [deployApi](https://github.com/Agoric/dapp-fungible-faucet/blob/main/api/deploy.js#L35), which is exported, should close over ambient access to `fs.promises.writeFile`. It seems straightforward to pass `writeFile` (or `writeDefaultFile`) in, just like `{ bundleSource, pathResolve }` are passed in.

**Comment by erights** [comment]
Hi @dckc so far you've been our most effective conscience getting us to clean up our act on this. Ok if I assign it to you? Not that I'm asking you to do this but rather to get people to do this for the code they're responsible for.

**Comment by dckc** [comment]
If you'll clarify the "finish line" a bit, I'm fine if you assign it to me.

Above I proposed "do not export powerful objects"; is that a good target / finish line?

**Comment by erights** [comment]
If someone is just using an ambient `fs` power to do something narrow, then they aren't *exporting* anything powerful. But the least authority they seem to require is immensely more that the least authority they actually require. Also, they should generally accept authority by parameters, not by ambient use. You know all this and have already been reacting on all these dimensions. Please keep doing so.

The finish line is when you've sufficiently habituated everyone else that we are all sensitive to it and reliably complain about it in code reviews. I don't know how to fit that into the goal of closing an issue, but it is what we must do. Adding @rowgraus to the assignees for that process issue. Suggestion: you coordinate with @rowgraus . You get all this started until he feels like he can take the process forward from there.

@dckc @rowgraus reasonable?

**Comment by dckc** [comment]
The target you're after is reasonably clear, @erights . I'm ok to aim for it.

@rowgraus you're welcome to help, but it seems a bit of an odd match for your role, to me.

**Comment by erights** [comment]
> @rowgraus you're welcome to help, but it seems a bit of an odd match for your role, to me.

I leave that to the two of you.

**Comment by erights** [comment]
We really need to do an inventory of all the uses of ambient authority scattered through our code. Since this bug was filed, it seems the issue has only been getting worse, not better. For one, just redoing the query that started this thread, we've gone from 33 scattered uses of `from 'fs';` to 39.

![Screen Shot 2021-03-07 at 12 23 54 PM](https://user-images.githubusercontent.com/273868/110253599-10b71300-7f40-11eb-931a-fcc37efdafa8.png)

**Comment by erights** [comment]
While I've been worrying about and asking for feedback on how to thread configuration options to `lockdown` https://github.com/Agoric/SES-shim/issues/578 , it seems we've been pervasively ignoring the underlying hazard and just passing environment variables around everywhere. In fact, it is worse. The following screenshot is of uses of `process.env`. All of these use `process` as a global variable. None of these have anything like a `/* global process */` declaration at the top of the file. Despite this absence, none of their uses of `process` is shown in vscode with a red squiggle underline and none generates a warning. In fact, the only such declaration I could find was in a test file, which was therefore excluded from that search. test-node-version.js begins with the peculiar text

```js
// eslint-disable-next-line no-redeclare
/* global process */
```

indicating that eslint considers this declaration a redeclaration. Where is the base declaration and how do we turn it off?

![Screen Shot 2021-03-07 at 12 29 57 PM](https://user-images.githubusercontent.com/273868/110253751-df8b1280-7f40-11eb-9569-5bf67e476690.png)

**Comment by erights** [comment]
One lesson I learned at Google is that an organization is much more likely to make incremental progress to a goal if it can turn it into a metric. We've been planning for a closer integration with LavaMoat anyway. It seems we need to do this sooner rather than later so that we can track all of these proliferating uses of ambient authority, understand where they all are, take inventory, identify the worst offenders to fix quickly, and then start making incremental but rapid progress on reducing the remainder.

**Comment by erights** [comment]
Hi @kumavis we need to start integrating LavaMoat into our own development process. In its absence, we've let uses of ambient authority proliferate.

**Comment by erights** [comment]
This concern is already on our https://github.com/Agoric/agoric-sdk/wiki/Code-Review-Checklist . We need to start applying it.

**Comment by erights** [comment]
Btw, I have made this mistake too. fatal-assert.js uses `process` as an ambient global, and a source of much more dangerous authority than `process.env` --- `process.abort` and `process.exit`. And it has no `/* global` comment or anything else to flag WARNING WARNING DANGER DANGER THERE BE DRAGONS HERE. Without tool support or any best practices around this issue, we'll all be too sloppy.

**Comment by erights** [comment]
![Screen Shot 2021-03-07 at 1 09 37 PM](https://user-images.githubusercontent.com/273868/110254985-6b536d80-7f46-11eb-830d-e8b7271aecbe.png)

**Comment by michaelfig** [comment]
I recognize the danger of ambient authority.  IIUC, many uses of it just need to be moved up to the script main entrypoint and propagated explicitly to other modules.

We should have conventions for that entrypoint's name, such as `rollup.config.js`, `packages/*/test/**/test-*.js`, `packages/*/scripts/*`, and `packages/*/bin/*`.  If you  ignore those patterns, how many uses are left?

**Comment by dtribble** [comment]
I see lots of things that are tools and command line widgets. So they generally need the authority that they are using, and we just need a way to lock it down.  Is that "load with LavaMoat" and then they just get the authority they are currently using?  Or do we need a library so that they are handed less authority (which would certainly be stronger).  For example, the program moves into a compartment and is just handed authority it needs, and then we have a fairly generic wrapper that captures the key authorities we intend to hand it.

**Comment by erights** [comment]
> I recognize the danger of ambient authority. IIUC, many uses of it just need to be moved up to the script main entrypoint and propagated explicitly to other modules.

Yes. That argument passing makes it explicit and virtualizable. In E this was the distinction between a tiny *.e files that functioned as a "main", and *.emaker files, that functioned as modules. We are set up to make the same distinction using compartments. Only a tiny "main" file, distinguished as you say by an unambiguous file name pattern, should ever execute in the start compartment. *Everything* else, including the SwingSet kernel, should be in compartments created by that "main". The "main" is responsible for setting up the top level compartments, wiring them together, and handing out least authority divisions and attenuations of authority available in the start compartment.

Once we are pervasively using LavaMoat analysis and policy files, some of this will change. But that's where we should start from.

> 
> We should have conventions for that entrypoint's name, such as `rollup.config.js`, `packages/*/test/**/test-*.js`, `packages/*/scripts/*`, and `packages/*/bin/*`. If you ignore those patterns, how many uses are left?

Well, the search went from "150 in 44 files" to "150 in 41 files" which does not make a lot of sense. But in any case, here it is. Easy enough to try it yourself of course.

Part 1 of 2 (notice the scroll bar)

![Screen Shot 2021-03-07 at 4 09 54 PM](https://user-images.githubusercontent.com/273868/110260390-e590eb80-7f60-11eb-9dc5-c674859a10fa.png)

Part 2 of 2 (notice the scroll bar)

![Screen Shot 2021-03-07 at 4 10 25 PM](https://user-images.githubusercontent.com/273868/110260394-eb86cc80-7f60-11eb-8842-c6f1aa27abf7.png)

**Comment by erights** [comment]
> I see lots of things that are tools and command line widgets. So they generally need the authority that they are using, and we just need a way to lock it down. Is that "load with LavaMoat" and then they just get the authority they are currently using? Or do we need a library so that they are handed less authority (which would certainly be stronger). For example, the program moves into a compartment and is just handed authority it needs, and then we have a fairly generic wrapper that captures the key authorities we intend to hand it.

Those are the right questions. Let's start there.

**Comment by erights** [comment]
We need a clear and unambiguous distinction between a "main" that should execute in a scope with lots of authority, because it is the logic for dividing and attenuation it, vs everything else.

We need to change our eslint settings so it is no longer treating, for example `process`, as a global that can be used without declaration and without warning. I could not figure out where this setting is.

**Comment by dckc** [comment]
> ... We need to change our eslint settings so it is no longer treating, for example `process`, as a global that can be used without declaration and without warning. I could not figure out where this setting is.

It's from "node - Node.js global variables and Node.js scoping." in [Language Options \- ESLint](https://eslint.org/docs/user-guide/configuring/language-options#specifying-environments).

Our rules inherit...

```js
  env: {
    node: true
  },
```
https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/node.js#L2-L4

This turns it off...

```diff
$ git diff
diff --git a/packages/eslint-config/eslint-config.json b/packages/eslint-config/eslint-config.json
index f781d7a2..fc31ec40 100644
--- a/packages/eslint-config/eslint-config.json
+++ b/packages/eslint-config/eslint-config.json
@@ -7,6 +7,7 @@
   ],
   "parser": "@typescript-eslint/parser",
   "env": {
+    "node": false,
     "es6": true
   },
   "globals": {
```

but then others such as `console` and `TextEncoder` get red squiggles too. We can turn those on by default...

```diff
$ git diff packages/eslint-config/
diff --git a/packages/eslint-config/eslint-config.json b/packages/eslint-config/eslint-config.json
index f781d7a2..9fdc5be1 100644
--- a/packages/eslint-config/eslint-config.json
+++ b/packages/eslint-config/eslint-config.json
@@ -7,10 +7,14 @@
   ],
   "parser": "@typescript-eslint/parser",
   "env": {
+    "node": false,
     "es6": true
   },
   "globals": {
     "globalThis": "readonly",
+    "console": "readonly",
+    "TextEncoder": "readonly",
+    "TextDecoder": "readonly",
     "BigInt": "readonly",
     "assert": "readonly",
     "harden": "readonly",
```

**Comment by erights** [comment]
Hi @dckc , sorry I just saw that. Already done all of that in #2594 , but thanks!

**Comment by dckc** [comment]
@erights on using LavaMoat... isn't that a way of managing legacy access to ambient authority? Why should we create more legacy? Why should _any_ of our packages export functions that close over powerful globals?

_p.s. one reason: shims that look at `globalThis` to see if the environment already provides, say, `HandledPromise`_

**Comment by dckc** [comment]
> We need a clear and unambiguous distinction between a "main" that should execute in a scope with lots of authority, because it is the logic for dividing and attenuation it, vs everything else.

Yes... `#!/usr/bin/env node` seems like a conventional way to signal a main script.

**Comment by dckc** [comment]
by way of example, I just refactored `ava-xs` to use a `#!/usr/bin/env node` script rather than a shell script:
https://github.com/Agoric/agoric-sdk/pull/2573/commits/fce51bccb292cbcdc3b42fbd58d5e436a974ae87

**Comment by erights** [comment]
> Yes... `#!/usr/bin/env node` seems like a conventional way to signal a main script.

I like that a lot!

**Comment by erights** [comment]
>  Why should we create more legacy? Why should any of our packages export functions that close over powerful globals?

We should not. Let's make progress towards not doing so. There will be some exceptions, but these should be understood as special cases that bear closer examination. 

Now that agoric-sdk mentions each of these global variables in a `/* global ... */` comment near the top of each file, we can take inventory are start to trim this back.

Aside from global variables, we also need to classify imports into two categories so we can take inventory of static excess authority there.

**Comment by michaelfig** [comment]
What LavaMoat can do is compartmentalise NPM packages.  We still do use NPM and this is a useful feature to have without having to rewrite everything we depend on (especially in the build process).

**Comment by erights** [comment]
Yes. For code we didn't write, much of the point of SES, Compartments, enablements, LavaMoat, and all the rest is exactly to support code we didn't write, while still constraining it to POLA. I took @dckc 's question to be only about code that we entirely write. Our code should mostly be in Jessie anyway, in which such static sources of authority should be impossible.

**Comment by dckc** [comment]
@warner writes in https://github.com/Agoric/agoric-sdk/pull/3207#issuecomment-850676448 :
>  I don't know where to draw the line between "this module expects to run in the start compartment and can thus `import` or `globalThis.foo` anything it needs", and "this module is pure and the caller must do `import`/etc and give it the following N things".

A "main" that should execute in a scope with lots of authority is indicated by `#!/usr/bin/env node` at the top (and +x file permissions). A test script is named `test-*.js`. No modules should export powerful objects (for example, objects that close over `fs`). Note that tests don't export anything.

I'm not sure to what extent the team has adopted this rule, but @erights writes Mar 7 in https://github.com/Agoric/agoric-sdk/issues/2160#issuecomment-792350511 :

> This concern is already on our https://github.com/Agoric/agoric-sdk/wiki/Code-Review-Checklist . We need to start applying it.

He also notes lack of tool support in https://github.com/Agoric/agoric-sdk/issues/2160#issuecomment-792352040 and utility of metrics to track progress in https://github.com/Agoric/agoric-sdk/issues/2160#issuecomment-792349015

I hope to try out lavamoat to establish a metric.

**Comment by abebeos** [comment]
> Strictly speaking, I don't think this constitutes a **bug**,

Personally, I like labeling such issues as `flaw` (as e.g. in `design-flaw`, `implementation-flaw` etc.). Label color same with `bug` label color.

**Comment by abebeos** [comment]
# Distressing proliferation of ambient authority

I have the feeling that this title  is "too much" for an IT project issue-tracker. Let's see:

* https://www.merriam-webster.com/dictionary/distressing
* https://www.merriam-webster.com/dictionary/proliferation
* https://www.merriam-webster.com/dictionary/ambient
* https://www.merriam-webster.com/dictionary/authority

=> Failing to extract the essence

Looking at https://people.eecs.berkeley.edu/~daw/talks/PLAS06.ppt, till page 4

> Authority: ability to affect the rest of the world.
Ambient authority: authority that’s available
even if you don’t ask for it.

Now I understand (and additionally conclude that I dislike the "ambient authority" terminology, mainly because of "ambient").

Trying to normalize:

=> Reduce Access to Potentially Dangerous Subsystems

Having understood the essence, I notice that this issue is additionally about introducing metrics and prevention/alerting mechanisms (so the devs realize their doing something wrong).

Issue is too dense, needs to be broken down into actionable sub-tasks/issues.

(me literally dizzy now, and not even reached my 1 hour limit )

**Comment by dckc** [comment]
An interesting tool to look for global mutable state is Moddable's `mutabilities`:
https://github.com/agoric-labs/xsnap-pub/blob/master/xsnap/documentation/lockdown.md#mutabilitieso
https://github.com/agoric-labs/xsnap-pub/tree/master/xsnap/examples/lockdown [3fc543d](https://github.com/agoric-labs/xsnap-pub/commit/3fc543d01a38c527b0e8edee441dfaea16e42083)
[on Mar 22, 2022](https://github.com/agoric-labs/xsnap-pub/commit/3fc543d01a38c527b0e8edee441dfaea16e42083)

It's currently commented out, but that's easy to address if you want to play around with it a bit:

```diff
diff --git a/xsnap/sources/xsnap.c b/xsnap/sources/xsnap.c
index 6389db2..4589083 100644
--- a/xsnap/sources/xsnap.c
+++ b/xsnap/sources/xsnap.c
@@ -375,14 +375,14 @@ void xsBuildAgent(xsMachine* machine)
        modInstallTextEncoder(the);
        modInstallBase64(the);
 //     
-//     xsResult = xsNewHostFunction(fx_harden, 1);
-//     xsDefine(xsGlobal, xsID("harden"), xsResult, xsDontEnum);
-//     xsResult = xsNewHostFunction(xs_lockdown, 0);
-//     xsDefine(xsGlobal, xsID("lockdown"), xsResult, xsDontEnum);
-//     xsResult = xsNewHostFunction(fx_petrify, 1);
-//     xsDefine(xsGlobal, xsID("petrify"), xsResult, xsDontEnum);
-//     xsResult = xsNewHostFunction(fx_mutabilities, 1);
-//     xsDefine(xsGlobal, xsID("mutabilities"), xsResult, xsDontEnum);
+       xsResult = xsNewHostFunction(fx_harden, 1);
+       xsDefine(xsGlobal, xsID("harden"), xsResult, xsDontEnum);
+       xsResult = xsNewHostFunction(xs_lockdown, 0);
+       xsDefine(xsGlobal, xsID("lockdown"), xsResult, xsDontEnum);
+       xsResult = xsNewHostFunction(fx_petrify, 1);
+       xsDefine(xsGlobal, xsID("petrify"), xsResult, xsDontEnum);
+       xsResult = xsNewHostFunction(fx_mutabilities, 1);
+       xsDefine(xsGlobal, xsID("mutabilities"), xsResult, xsDontEnum);
 
        xsEndHost(machine);
 }
```

**Comment by kumavis** [comment]
could make an eslint rule that gets angry when you import builtin modules

**Comment by turadg** [comment]
> could make an eslint rule that gets angry when you import builtin modules

+:100: to automating whatever the policy is. We don't even need a new rule if we're content with listing the most problematic: https://eslint.org/docs/latest/rules/no-restricted-imports

Then we can use the tooling to partition paths in which is really is a problem to import these from those were it's not. E.g. CLI tools.

**Comment by dckc** [comment]
LavaMoat-like techniques are starting to look feasible: https://github.com/LavaMoat/LavaMoat/issues/228#issuecomment-1486624429

here's hoping for time to dig in more...
